src/main/java/com/aws/greengrass/mqtt/bridge/MessageBridge.java

```diff
@@ -146,8 +156,21 @@ private <T extends Message> void handleMessage(TopicMapping.TopicType sourceType
     private <T extends Message> void publishMessage(MessageClient<T> client, String topic, Message message)
             throws MessageClientException {
         T msg = client.convertMessage(message);
-        msg = (T) msg.newFromMessageWithTopic(topic);
-        client.publish(msg);
+
+        if (isRetainAsPublished(topic) && msg instanceof MqttMessage) {
+            MqttMessage mqttMsg = (MqttMessage) msg;
+            mqttMsg = mqttMsg.toBuilder().retain(mqttMsg.isRetain()).build();
```

Comment:
you're reading from the same message that you're writing into. I don't think this is doing anything.

Reply:
I'm updating the logic here to check if the route options have retain as published set to false. If they do, I can bridge the message without retain, otherwise I'll bridge it as is